ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -799,6 +807,52 @@ private void writeMessagesToLocalStoreAndAdvanceTokens(List<ExchangeMetadataResp
         replicatingFromRemoteColo, replicatingOverSsl, datacenterName);
   }
 
+  /**
+   * Applies a TTL update to the blob described by {@code messageInfo}.
+   * @param messageInfo the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   * @throws IOException
+   * @throws MessageFormatException
+   * @throws StoreException
+   */
+  private void applyTtlUpdate(MessageInfo messageInfo, RemoteReplicaInfo remoteReplicaInfo)
+      throws IOException, MessageFormatException, StoreException {
+    MessageFormatInputStream ttlUpdateStream =
+        new TtlUpdateMessageFormatInputStream(messageInfo.getStoreKey(), messageInfo.getAccountId(),
+            messageInfo.getContainerId(), messageInfo.getOperationTimeMs());
+    MessageInfo info =
+        new MessageInfo(messageInfo.getStoreKey(), ttlUpdateStream.getSize(), false, true, messageInfo.getAccountId(),
+            messageInfo.getContainerId(), messageInfo.getOperationTimeMs());
+    MessageFormatWriteSet writeSet = new MessageFormatWriteSet(ttlUpdateStream, Collections.singletonList(info), false);
+    DataNodeId remoteNode = remoteReplicaInfo.getReplicaId().getDataNodeId();
+    try {
+      // NOTE: It is possible that the key in question may have expired and this TTL update is being applied after it
+      // is deemed expired. The store will accept the op (BlobStore looks at whether the op was valid to do at the time
+      // of the op, not current time) but if compaction is running at the same time and has decided to clean up the
+      // record before this ttl update was applied (and this didn't find the key missing because compaction has not yet
+      // committed), then we have a bad situation where only a TTL update exists in the store. This problem has to be
```

Comment:
Yup - it will be removed because there is no "anchoring" put.

Reply:
A stale TTL update means that the blob is permanent in other stores though, right? What is the stand on that? That it will only happen in rare scenarios?

Reply:
Yes - will happen only on rare scenarios and is a corner case that I need to give some thought to (probably find a way to delay compaction if the store is "too behind).